rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/CharacteristicLongWriteOperation.java

```diff
@@ -209,4 +214,58 @@ public Boolean call(Object emission) {
             }
         };
     }
+
+    private static Func1<Observable<? extends Throwable>, Observable<?>> retryOperationStrategy(
+            final WriteOperationRetryStrategy writeOperationRetryStrategy,
+            final ByteBuffer byteBuffer,
+            final int batchSize) {
+        return new Func1<Observable<? extends Throwable>, Observable<?>>() {
+            @Override
+            public Observable<?> call(Observable<? extends Throwable> emittedOnWriteFailure) {
+                return emittedOnWriteFailure
+                        .flatMap(canRetryError())
+                        .map(toLongWriteFailureObject())
+                        .compose(writeOperationRetryStrategy)
+                        .doOnNext(replaceByteBufferPositionForRetry());
+            }
+
+            @NonNull
+            private Func1<Throwable, WriteOperationRetryStrategy.LongWriteFailure> toLongWriteFailureObject() {
+                return new Func1<Throwable, WriteOperationRetryStrategy.LongWriteFailure>() {
+                    @Override
+                    public WriteOperationRetryStrategy.LongWriteFailure call(Throwable throwable) {
+                        final int failedBatchNumber = (byteBuffer.position() - batchSize) / batchSize;
+                        return new WriteOperationRetryStrategy.LongWriteFailure(
+                                failedBatchNumber,
+                                (BleGattCharacteristicException) throwable
+                        );
+                    }
+                };
+            }
+
+            @NonNull
+            private Action1<WriteOperationRetryStrategy.LongWriteFailure> replaceByteBufferPositionForRetry() {
+                return new Action1<WriteOperationRetryStrategy.LongWriteFailure>() {
+                    @Override
+                    public void call(WriteOperationRetryStrategy.LongWriteFailure longWriteFailure) {
+                        final int newBufferPosition = longWriteFailure.getBatchNumber() * batchSize;
+                        byteBuffer.position(newBufferPosition);
+                    }
+                };
+            }
+
+            @NonNull
+            private Func1<Throwable, Observable<Throwable>> canRetryError() {
```

Comment:
Maybe `toLongWriteFailureOrError()`? It would then look like `.flatMap(toLongWriteFailureOrError())`

Reply:
You mean merge those two?
```
.flatMap(canRetryError())
.map(toLongWriteFailureObject())
```

Reply:
Yes, as the the `.map()` contains a pretty straightforward logic. Maybe `toLongWriteFailureIfRetryable()` (not sure about the spelling)

Reply:
I've split them based on comment: https://github.com/Polidea/RxAndroidBle/pull/357#discussion_r162981644 :)
Haha actually, I was wondering about the same thing, maybe `toLongWriteFailureIfCanBeRetried()`?

Reply:
It was exactly more about using `.compose(writeOperationRetryStrategy)`

Reply:
Ok :)